frameworks/PHP/hamlet/swoole.php

```diff
@@ -15,6 +15,6 @@
 );
 $application = new Application($database);
 $onWorkerStart = function () use (&$database) {
-    $database->warmUp((int) (768 / swoole_cpu_num()));
+    $database->warmUp((int) (1024 / swoole_cpu_num()));
```

Comment:
The number of connections to the database doesn't change anything. You only need 512 db connections.
If you have failed transactions, it is because the Http server does not respond to some requests, and you'll have to determine why.

Reply:
And yet it helps for some reason. My impression is that some of them don't return to the pool either on time or ever. Hopefully I'll get more time to take a closer look at this in the following weeks.

Reply:
In principle:
For `Query` or `Update` Test:
- I get a connection from the pool.
- I make the requests with (20 select / update)
- I put the connection back into the pool.
That's why I only need 512 connections.
Hamlet seems to work differently, **Hamlet** seems to work differently, since I don't see in your test code any access to the pool.

Reply:
The pool code is in the https://github.com/hamlet-framework/db-mysql-swoole

The idea of `Database` abstraction is to keep the creation of new connections, transactions, exception handling and pooling inside of this class, so that database access code stays the same between swoole and no-swoole environments.

Reply:
It's a nice goal,
but when (or where) are you getting a new connection from the pool (which part of the code)?

Reply:
https://github.com/hamlet-framework/db/blob/fbff8ec3b8fb9c254edec61e090beeb349741a96/src/Database.php#L86

Reply:
so there's a fetch from the pool for each request?

Reply:
It will call `$pool->pop()` 20 times. Do you think this is the issue?

Reply:
The alternative would be to expose pool management directives to the caller, something like 
```
$database->withSameConnection(function () {
     // do 20 calls within
});
```
not sure I like it, as I am not sure what's the use case of it outside of benchmarking.

Reply:
> It will call `$pool->pop()` 20 times. Do you think this is the issue?

Yes, it's huge.
That's why it works with `Fortunes` and `Db` tests, and not with the others.
You would need `512*20` ready connections in your pool => 10240

Reply:
They executors release connections back into the pool immediately after, so should the 512*20 be the upper boundary?

Reply:
I think so, yes.
so `512*20/worker_count` to create for each worker...
you can't control concurrent accesses, especially in asynchronous mode.
But anyway, it's too many connections (too many resources), and too many manipulations on the pool to be performant.
A blocking PDO connection would be more efficient in this case.

Reply:
Can you confirm, even with 1024 connections you still see 5xx when running 
```
./tfb --mode benchmark --test hamlet-swoole --type update
```

Reply:
Confirmation with 1024 db connexions:
The results are unpredictable.

![image](https://user-images.githubusercontent.com/2511052/67942313-d913ba80-fbd7-11e9-91bb-6e3313519876.png)

![image](https://user-images.githubusercontent.com/2511052/67942368-fa74a680-fbd7-11e9-9370-a089618bcc97.png)